compiler/src/iree/compiler/Codegen/Common/PadDynamicAlloc.cpp

```diff
@@ -37,22 +40,47 @@ static Value skipAffineMaxZero(Value dim) {
   return *affineMax.getSymbolOperands().begin();
 }
 
-static LogicalResult padAlloc(MLIRContext *context, memref::AllocOp allocOp) {
+static FailureOr<int64_t> getUpperBound(Value dim,
+                                        const DataFlowSolver &solver) {
+  // Check the integer range analysis.
+  if (auto *maybeRange =
+          solver.lookupState<dataflow::IntegerValueRangeLattice>(dim)) {
+    IntegerValueRange range = maybeRange->getValue();
+    if (!range.isUninitialized() &&
+        range.getValue().smax() !=
+            IntegerValueRange::getMaxRange(dim).getValue().smax()) {
+      return range.getValue().smax().getSExtValue();
+    }
+  }
+
+  // Check the value bounds constraint set.
+  // TODO: These two analysis could be merged, but probably needs
+  // to happen usptream.
+  auto ub = ValueBoundsConstraintSet::computeConstantBound(
+      presburger::BoundType::UB, {dim, /*dim=*/std::nullopt},
+      /*stopCondition=*/nullptr, /*closedUB=*/true);
+  if (succeeded(ub)) {
+    return ub.value();
+  }
+  return failure();
```

Comment:
`return ub;`?

Reply:
I dont know what happen really when I return ub cause it is failure. That is hard for me to follow. This seems more explicit.